clib/cEphemeron.ml

```diff
@@ -8,84 +8,104 @@
+type boxkey = Box : 'a typ -> boxkey [@@unboxed]
+
+(* Carry the type we just erased with the actual key. *)
+type 'a key = 'a typ * boxkey
+
+module EHashtbl = Ephemeron.K1.Make(struct
+  type t = boxkey
+  let equal = (==)
+  let hash  = Hashtbl.hash
 end)
 
-(* A key is the (unique) value inside a boxed key, hence it does not
-   keep its corresponding boxed key reachable (replacing key_type by boxed_key
-   would make the key always reachable) *)
-let values : Obj.t HT.t = HT.create 1001
-
-(* To avoid a race condition between the finalization function and
-   get/create on the values hashtable, the finalization function just
-   enqueues in an imperative list the item to be collected.  Being the list
-   imperative, even if the Gc enqueues an item while run_collection is operating,
-   the tail of the list is eventually set to Empty on completion.
-   Kudos to the authors of Why3 that came up with this solution for their
-   implementation of weak hash tables! *)
-type imperative_list = cell ref
-and cell = Empty | Item of key_type ref * imperative_list
-
-let collection_queue : imperative_list ref = ref (ref Empty)
-
-let enqueue x = collection_queue := ref (Item (!x, !collection_queue))
-
-let run_collection () =
-  let rec aux l = match !l with
-    | Empty -> ()
-    | Item (k, tl) -> HT.remove values k; aux tl in
-  let l = !collection_queue in
-  aux l;
-  l := Empty
-
-(* The only reference to the boxed key is the one returned, when the user drops
-   it the value eventually disappears from the values table above *)
-let create (v : 'a) : 'a key =
-  run_collection ();
-  let k = mk_key () in
-  HT.add values !k (Obj.repr v);
-  Gc.finalise enqueue k;
-  k
+type value = { get : 'k. 'k typ -> 'k } [@@unboxed]
+
+let values : value EHashtbl.t =
+  EHashtbl.create 1001
+
+let create : type v. v -> v key =
+  fun value ->
+  let module M = struct
+    type _ typ += Typ : v typ
+
+    let get : type k. k typ -> k =
+      fun typ ->
+      match typ with
+      | Typ -> value
+      | _ -> assert false
+
+    let boxkey = Box Typ
+    let key    = Typ, boxkey
+    let value  = { get }
+  end in
+  EHashtbl.add values M.boxkey M.value;
+  M.key
 
 (* Avoid raising Not_found *)
 exception InvalidKey
-let get (k : 'a key) : 'a =
-  run_collection ();
-  try Obj.obj (HT.find values !k)
+let get (typ, boxkey) =
+  try (EHashtbl.find values boxkey).get typ
   with Not_found -> raise InvalidKey
 
-(* Simple utils *)
-let default k v =
-  try get k
-  with InvalidKey -> v
+let default (typ, boxkey) default =
+  try (EHashtbl.find values boxkey).get typ
+  with Not_found -> default
 
-let iter_opt k f =
-  match
-    try Some (get k)
-    with InvalidKey -> None
-  with
-  | None -> ()
-  | Some v -> f v
+let iter_opt (typ, boxkey) f =
+  try f ((EHashtbl.find values boxkey).get typ)
+  with Not_found -> ()
 
-let clear () = run_collection ()
+(* NB: clear only removes dead bindings, so it uses Ephemeron.S.clean *)
```

Comment:
Can you explain this comment? I've hard time understand what it means.
The semantics of the previous clear was to force a collection, that looks like "only removes dead bindings".

Reply:
A function `clear` calling a function `EHashtbl.clean` when `EHashtbl.clear` also exists looks like a typo. Here it is not a typo, the semantics of `clear` correspond to `EHashtbl.clean`. Maybe the `CEphemeron.clear` should just be renamed.

Reply:
Renamed.